# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1680722105 -7200
# Parent  58c0dfead17ac02c88874d3bcb0bc4c772b422e1
Feature: When multiple message are selected, current message is shown.

Vaguely inspired by Mats Palmgren's work from https://bugzilla.mozilla.org/show_bug.cgi?id=364896

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -1336,8 +1336,10 @@ pref("app.donation.eoy.version.viewed", 
 // Use our own prefs for the end-of-year appeal. Otherwise running TB on the
 // profile would show the TB appeal and then never again the BB appeal.
 pref("betterbird.donation.eoy.version", 2022);
 pref("betterbird.donation.eoy.version.viewed", 0);
 pref("betterbird.donation.eoy.url", "https://www.betterbird.eu/end-of-year/");
 
 // Make this visible in the advanced preferences.
 pref("ui.prefersReducedMotion", 0);
+
+pref("mail.multiselect_message_shows_current", false);
diff --git a/mail/base/content/folderDisplay.js b/mail/base/content/folderDisplay.js
--- a/mail/base/content/folderDisplay.js
+++ b/mail/base/content/folderDisplay.js
@@ -229,16 +229,17 @@ FolderDisplayWidget.prototype = {
    * @return true if the selection should be summarized for this folder. This
    *     is based on the mail.operate_on_msgs_in_collapsed_threads pref and
    *     if we are in a newsgroup folder. XXX When bug 478167 is fixed, this
    *     should be limited to being disabled for newsgroups that are not stored
    *     offline.
    */
   get summarizeSelectionInFolder() {
     return (
+      !Services.prefs.getBoolPref("mail.multiselect_message_shows_current") &&
       Services.prefs.getBoolPref("mail.operate_on_msgs_in_collapsed_threads") &&
       !(this.displayedFolder instanceof Ci.nsIMsgNewsFolder)
     );
   },
 
   /**
    * @return the nsITreeSelection object for our tree view.  This exists for
    *     the benefit of message tabs that haven't been switched to yet.
diff --git a/mailnews/base/src/nsMsgDBView.cpp b/mailnews/base/src/nsMsgDBView.cpp
--- a/mailnews/base/src/nsMsgDBView.cpp
+++ b/mailnews/base/src/nsMsgDBView.cpp
@@ -1220,30 +1220,29 @@ nsMsgDBView::SelectionChangedXPCOM() {
       mSummarizeFailed = true;
     }
   }
 
   bool summaryStateChanged = selectionSummarized != mSelectionSummarized;
   mSelectionSummarized = selectionSummarized;
 
   // If only one item is selected then we want to display a message.
-  if (mTreeSelection && selection.Length() == 1 && !selectionSummarized) {
-    int32_t startRange;
-    int32_t endRange;
-    nsresult rv = mTreeSelection->GetRangeAt(0, &startRange, &endRange);
+  // If we're not summarized, we display the current message.
+  if (mTreeSelection && selection.Length() >= 1 && !selectionSummarized) {
+    int32_t currentIndex;
+    nsresult rv = mTreeSelection->GetCurrentIndex(&currentIndex);
     // Tree doesn't care if we failed.
     NS_ENSURE_SUCCESS(rv, NS_OK);
 
-    if (startRange >= 0 && startRange == endRange &&
-        uint32_t(startRange) < GetSize()) {
+    if (currentIndex >= 0 && uint32_t(currentIndex) < GetSize()) {
       if (!mRemovingRow) {
         if (!mSuppressMsgDisplay)
-          LoadMessageByViewIndex(startRange);
+          LoadMessageByViewIndex(currentIndex);
         else
-          UpdateDisplayMessage(startRange);
+          UpdateDisplayMessage(currentIndex);
       }
     } else {
       // Selection seems bogus, so set to 0.
       selection.Clear();
     }
   } else {
     // If we have zero or multiple items selected, we shouldn't be displaying
     // any message.
